File: src/main/java/link/infra/packwiz/installer/request/handlers/RequestHandlerHTTP.java
package link.infra.packwiz.installer.request.handlers;

import link.infra.packwiz.installer.metadata.SpaceSafeURI;
import link.infra.packwiz.installer.request.IRequestHandler;
import okio.Okio;
import okio.Source;

import java.net.URLConnection;

public class RequestHandlerHTTP implements IRequestHandler {

	@Override
	public boolean matchesHandler(SpaceSafeURI loc) {
		String scheme = loc.getScheme();
		return "http".equals(scheme) || "https".equals(scheme);
	}

	@Override
	public Source getFileSource(SpaceSafeURI loc) throws Exception {
		URLConnection conn = loc.toURL().openConnection();
		// TODO: when do we send specific headers??? should there be a way to signal this?
		// github *sometimes* requires it, sometimes not!
		//conn.addRequestProperty("Accept", "application/octet-stream");
		// 30 second read timeout
		conn.setReadTimeout(30 * 1000);
		return Okio.source(conn.getInputStream());
	}

}

File: src/main/kotlin/link/infra/packwiz/installer/request/handlers/RequestHandlerHTTP.kt
package link.infra.packwiz.installer.request.handlers

import link.infra.packwiz.installer.request.IRequestHandler
import okhttp3.HttpUrl
import okhttp3.OkHttpClient
import okhttp3.Request
import okio.IOException
import okio.Source
import okio.source
import java.net.HttpURLConnection
import java.util.concurrent.TimeUnit

open class RequestHandlerHTTP : IRequestHandler {
	private val client: OkHttpClient

	init {
		client = OkHttpClient.Builder()
				// 30 second read timeout
				.readTimeout(30 * 1000, TimeUnit.MILLISECONDS)
				.build()
	}

	override fun matchesHandler(loc: HttpUrl): Boolean {
		val scheme = loc.scheme
		return "http" == scheme || "https" == scheme
	}

	override fun getFileSource(loc: HttpUrl): Source? {
		val request = Request.Builder()
				.url(loc)
				// TODO: when do we send specific headers??? should there be a way to signal this?
				.addHeader("Accept", "application/octet-stream")
				// TODO: include version?
				.addHeader("User-Agent", "packwiz-installer")

				.method("GET", null)
				.build()

		val response = client.newCall(request).execute()

		if (!response.isSuccessful) throw IOException("Unexpected code $response")

		return response.body!!.source()
	}
}